pkg/chunk/cassandra/storage_client.go

```diff
@@ -0,0 +1,219 @@
+package cassandra
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"time"
+
+	"github.com/gocql/gocql"
+
+	"github.com/weaveworks/cortex/pkg/chunk"
+)
+
+const (
+	maxRowReads = 100
+)
+
+// Config for a StorageClient
+type Config struct {
+	address           string
+	keyspace          string
+	replicationFactor int
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.address, "cassandra.address", "", "Address of Cassandra instances.")
+	f.StringVar(&cfg.keyspace, "cassandra.keyspace", "", "Keyspace to use in Cassandra.")
+	f.IntVar(&cfg.replicationFactor, "cassandra.replication-factor", 1, "Replication factor to use in Cassandra.")
+}
+
+func (cfg *Config) session() (*gocql.Session, error) {
+	if err := cfg.createKeyspace(); err != nil {
+		return nil, err
+	}
+
+	cluster := gocql.NewCluster(cfg.address)
+	cluster.Keyspace = cfg.keyspace
+	cluster.Consistency = gocql.Quorum
+	cluster.BatchObserver = observer{}
+	cluster.QueryObserver = observer{}
+
+	return cluster.CreateSession()
+}
+
+// createKeyspace will create the desired keyspace if it doesn't exist.
+func (cfg *Config) createKeyspace() error {
+	cluster := gocql.NewCluster(cfg.address)
+	cluster.Keyspace = "system"
+	cluster.Timeout = 20 * time.Second
+	session, err := cluster.CreateSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	return session.Query(fmt.Sprintf(
+		`CREATE KEYSPACE IF NOT EXISTS %s
+		 WITH replication = {
+			 'class' : 'SimpleStrategy',
+			 'replication_factor' : %d
+		 }`,
+		cfg.keyspace, cfg.replicationFactor)).Exec()
+}
+
+// storageClient implements chunk.storageClient for GCP.
+type storageClient struct {
+	cfg       Config
+	schemaCfg chunk.SchemaConfig
+	session   *gocql.Session
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	session, err := cfg.session()
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageClient{
+		cfg:       cfg,
+		schemaCfg: schemaCfg,
+		session:   session,
+	}, nil
+}
+
+func (s *storageClient) Close() {
+	s.session.Close()
+}
+
+type writeBatch struct {
+	b *gocql.Batch
+}
+
+func (s *storageClient) NewWriteBatch() chunk.WriteBatch {
+	return writeBatch{
+		b: gocql.NewBatch(gocql.UnloggedBatch),
+	}
+}
+
+func (b writeBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	b.b.Query(fmt.Sprintf("INSERT INTO %s (hash, range, value) VALUES (?, ?, ?)", tableName),
+		hashValue, rangeValue, value)
+}
+
+func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	cassandraBatch := batch.(writeBatch)
+	return s.session.ExecuteBatch(cassandraBatch.b.WithContext(ctx))
```

Comment:
Out of curiosity, would you know, for large volumes, how these batches behave?
If it is what I think it is, I'd expect the coordinator node to suffer a bit under large batches vs. sending writes directly to the replicas (does GoCQL support that at all?).

Also, do we need batches for a specific reason? (e.g. atomicity/isolation of the writes)

Reply:
There is no single coordinator node in Cassandra - GoCQL will round robin these amongst nodes (see RoundRobinHostPolicy). The batches are pretty small anyway (~10 keys IIRC).

> Also, do we need batches for a specific reason? (e.g. atomicity/isolation of the writes)

Nope, just picking most efficient thing.

Reply:
> _There is no single coordinator node in Cassandra_

Yes, absolutely.

Sorry if I wasn't clear, but I was referring to the fact one batch might contain data with different row keys, which eventually will have to be routed to their respective replica nodes by the coordinator node.
*If* this is indeed the case, and *if* the batch can end up being large, the work the coordinator will have to do might be non-negligible vs. removing the extra hop and sending the data directly to the replicas.

The `RoundRobinHostPolicy` will make sure "all nodes are equally swamped" when they take the role of the coordinator node, and therefore average out the load, but it might still be unnecessary load, if that makes sense.

To quote the documentation (in case that's clearer):

> A well constructed batch targeting a single partition can reduce client-server traffic and more efficiently update a table with a single row mutation. A batch can also target multiple partitions, when atomicity and isolation is required. Multi-partition batches may decrease throughput and increase latency. Only use a multi-partition batch when there is no other viable option, such as asynchronous statements.

Reply:
Yes, that is true - but I've gone for an "Unlogged", avoiding the overheads of the batch atomicity.

Using batches probably doesn't add anything here, and I guess there is a chance they are suboptimal.  All writes will have an extra hop any way, to write the second replica, and the default policy is to just round robin, which means writes will seldom hit a node containing one replica to begin with.

If this causes trouble I'll investigate token aware routing and removing the batching on this.

Reply:
:+1: